File: compute.py
import os

import cv2 as cv
import numpy as np


def compute_finger(images_path: str, clear_path: str, save_path: str) -> None:
    finger = []
    sift = cv.SIFT_create()
    clear: np.ndarray = cv.imread(clear_path, 0)
    for root, _, files in os.walk(images_path):
        for file in files:
            if file.endswith(".png"):
                image: np.ndarray = cv.imread(os.path.join(root, file),
                                              0) - clear

                minimum: int = np.min(image)
                maximum: int = np.max(image)
                tmp = 255 / (maximum - minimum)
                image = np.uint8(np.round(tmp * image - minimum * tmp))

                keypoints: tuple
                descriptors: np.ndarray
                keypoints, descriptors = sift.detectAndCompute(image, None)

                length = len(keypoints)
                data = np.empty((2, length), np.ndarray)
                for i in range(length):
                    data[0, i] = keypoints[i].pt
                    data[1, i] = descriptors[i]

                finger.append(data)

    np.savez_compressed(save_path, *finger)


if __name__ == "__main__":
    compute_finger('fingerprints/finger-0', 'fingerprints/clear.png',
                   'fingerprints/finger-0.npz')

File: cpp/compute.cpp
#include "structs.hpp"


string ext = ".png";
string dir = "/home/mango/fpr";

void compute_finger(string images_path, string clear_path, string save_path) {
	cv::Mat clear = cv::imread(clear_path, 0);

	int i = 0;
	for (const auto file : recursive_directory_iterator(images_path)) {
		string path = file.path();

		if ((path.substr(path.size() - 4, 4) == ext) && path != clear_path) {
			cv::Mat image = clear - cv::imread(path, 0);

			cv::Mat ROI = cv::Mat::ones(cv::Size(image.size[0], image.size[1]), 0);

			double maximum;
			double minimum;

			minMaxLoc(image, &minimum, &maximum, NULL, NULL);

			double tmp = 255 / (maximum - minimum);
			image = tmp * (image - minimum);

			vector<cv::KeyPoint> keypoints;
			cv::Mat descriptors;

			cv::SIFT::create()->detectAndCompute(image, ROI, keypoints, descriptors);

			cv::FileStorage store(
					save_path + to_string(i) + ".yml",
					cv::FileStorage::WRITE);
			cv::write(store, "keypoints", keypoints);
			cv::write(store, "descriptors", descriptors);
			store.release();
			i++;
		}
	}
}

int main() {
	compute_finger("/home/mango/fpr", "/home/mango/fpr/clear.jpg", "/home/mango/fpr");
}
